pkg/webhook/node/plugins/resourceamplification/resource_amplification.go

```diff
@@ -0,0 +1,162 @@
+		corev1.ResourceCPU,
+		corev1.ResourceMemory,
+	}
+)
+
+func NewPlugin() *NodeResourceAmplificationPlugin {
+	plugin := &NodeResourceAmplificationPlugin{}
+	return plugin
+}
+
+var _ plugins.NodePlugin = &NodeResourceAmplificationPlugin{}
+
+type NodeResourceAmplificationPlugin struct {
+}
+
+func (n *NodeResourceAmplificationPlugin) Name() string {
+	return "NodeResourceAmplificationPlugin"
+}
+
+func (n *NodeResourceAmplificationPlugin) Validate(ctx context.Context, req admission.Request, node, oldNode *corev1.Node) error {
+	return fmt.Errorf("not implemented")
+}
+
+// Admit makes an admission decision based on the request attributes
+func (n *NodeResourceAmplificationPlugin) Admit(ctx context.Context, req admission.Request, node, oldNode *corev1.Node) error {
+	klog.V(3).Infof("enter NodeResourceAmplificationPlugin plugin admit for %s", req.AdmissionRequest.Name)
+
+	op := req.AdmissionRequest.Operation
+	switch op {
+	case admissionv1.Create:
+		return nil
+	case admissionv1.Update:
+		return n.handleUpdate(oldNode, node)
+	}
+	return nil
+}
+
+// isSupportedResourceChanged checks if the supported resources are changed
+func (n *NodeResourceAmplificationPlugin) isSupportedResourceChanged(oldNode, node *corev1.Node) bool {
+	if oldNode == nil || node == nil {
+		return false
+	}
+	oldOriginalResources := corev1.ResourceList{}
+	newOriginalResources := corev1.ResourceList{}
+	for _, resourceName := range supportedResources {
+		oldResourceValue, ok := oldNode.Status.Allocatable[resourceName]
+		if ok {
+			oldOriginalResources[resourceName] = oldResourceValue
+		}
+		resourceValue, ok := node.Status.Allocatable[resourceName]
+		if ok {
+			newOriginalResources[resourceName] = resourceValue
+		}
+	}
+	return !util.IsResourceListEqual(oldOriginalResources, newOriginalResources)
+}
+
+func (n *NodeResourceAmplificationPlugin) handleUpdate(oldNode, node *corev1.Node) error {
+	if node.Annotations[extension.AnnotationNodeResourceAmplificationRatio] == "" {
+		// When the feature is turned off, the saved raw allocatable needs to be cleaned up
+		delete(node.Annotations, extension.AnnotationNodeRawAllocatable)
+		return nil
+	}
+	if node.Status.Allocatable == nil {
+		// do nothing
+		return nil
+	}
+
+	// Set original resource under two conditions
+	// 1. if the original resource is not set
+	// 2. Node allocatable are updated, such as the administrator adjusting the reserved resource
+	//    https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources
+	//    Here, we rely on an assumption that only kubelet will update allocatable, and other roles will only expand new fields and not overwrite native fields.
+	//    FIXME If the updated resource of kubelet happens to be the same as the amplificated allocatable, the update will not take effect.
+	originalResources := corev1.ResourceList{}
+	if !extension.HasNodeRawAllocatable(node.Annotations) || n.isSupportedResourceChanged(oldNode, node) {
```

Comment:
If the webhook is restarting or the webhook crashes, and the resources of some nodes change during the period, for example, the memory increases from 32GiB to 64GiB, then after the webhook restarts and recovers, the Node Update event obtained again may be that the allocatables of oldNode and newNode are consistent. In this scenario, according to the current implementation, the past 32GiB will be multiplied by the amplification ratio, and the result may be less than 64GiB. In the same similar scenario, 32GiB is adjusted to 16GiB. Perhaps we should consider here that if the amplified value is smaller than the current node.status.allocatable, we should update the raw allocatable with the node status and re-adjust the node.status.allocatable;

Reply:
Users can adjust the webhook fail policy, but this is currently not recommended, I added a comment for it.